hal/src/rtl872x/i2c_hal.cpp

```diff
@@ -189,26 +203,30 @@ class I2cClass {
         conf.set_value = true;
         conf.value = 1;
         hal_gpio_configure(sdaPin_, &conf, nullptr);
-        conf.value = hal_gpio_read(sclPin_);
-        hal_gpio_configure(sclPin_, &conf, nullptr);
 
-        // Generate up to 9 pulses on SCL to tell slave to release the bus
-        for (int i = 0; i < 9; i++) {
-            hal_gpio_write(sdaPin_, 1);
-            HAL_Delay_Microseconds(50);
+        // Check the SCL first
+        hal_gpio_mode(sclPin_, INPUT_PULLUP);
+        if (hal_gpio_read(sclPin_) == 0) {
```

Comment:
@XuGuohui Can you explain why we only perform the reset if SCL is low? I don't think this is right. SCL is driven by master unless clock-stretching is in play or multi-master setup. Even if we forget about those cases, idle SCL state is 1, not 0, so we've just completely bypassed the whole reset sequence.

Reply:
My understanding is that, the reset procedure is used when the  bus is scretched by slave, or for whatwever reason the bus is halted. If the SCL is high, the master can take full control of the bus and it can simply issue the stop signal, no?

Reply:
Nope, this isn't correct, SCL doesn't matter and will not get locked up in our use-cases unless there is a hardware issue of course. The main purpose of reset procedure (and what the test validates as well) is to cleanly terminate an I2C transaction which might have been terminated somewhere in the middle (e.g. master was reset) and to avoid writing garbage data into the slave and also of course let the slave release SDA because it's still waiting for some number of bits (hence clocking 9 bits at most). See https://www.analog.com/media/en/technical-documentation/application-notes/54305147357414AN686_0.pdf as a reference for a more verbose explanation.